dbt/adapters/bigquery/dataset.py

```diff
@@ -0,0 +1,35 @@
+import imp
+from typing import List
+from google.cloud.bigquery import DatasetReference, Dataset, AccessEntry
+
+from dbt.events import AdapterLogger
+
+logger = AdapterLogger("BigQuery")
+
+
+def add_access_entry_to_dataset(dataset: Dataset, access_entry: AccessEntry) -> Dataset:
+    """Idempotently adds an access entry to a dataset
+
+    Args:
+        dataset (Dataset): the dataset to be updated
+        access_entry (AccessEntry): the access entry to be added to the dataset
+
+    Returns:
+        Dataset
+    """
+    access_entries: List[AccessEntry] = dataset.access_entries
+    # we can't simply check if an access entry is in the list as the current equality check
+    # does not work because the locally created AccessEntry can have extra properties.
+    for existing_entry in access_entries:
+        role_match = existing_entry.role == access_entry.role
+        entity_type_match = existing_entry.entity_type == access_entry.entity_type
+        if role_match and entity_type_match:
+            new_prop = access_entry._properties
+            existing_prop = existing_entry._properties
+            properties_match = [new_prop[k] == v for k,v in existing_prop.items()]
+            if properties_match:
```

Comment:
Perhaps I'm missing something, but can we simplify lines 27-30:
```
if access_entry._properties == existing_entry._properties:
```

If so, we combine this with lines 24 and 25 as well:
```
for existing_entry in access_entries:
    if all(
        existing_entry.role == access_entry.role,
        existing_entry.entity_type == access_entry.entity_type,
        existing_entry._properties == access_entry._properties
    ):
        logger.warning(f"Access entry {access_entry} already exists in dataset")
        return dataset
```

And then if we want to be really fancy, we can define ```def __eq__(self)``` on AccessEntry, which turns this whole function into:
```
if access_entry in dataset.access_entries:
    logger.warning(f"Access entry {access_entry} already exists in dataset")
    return dataset
access_entries.append(access_entry)
dataset.access_entries = access_entries
return dataset
```
which can be simplified further to:
```
dataset.access_entries = list(set(dataset.access_entries).add(access_entry))
return dataset
```

And if we can change the datatype of dataset.access_entries, then this whole function can be avoided as it's equivalent to:
```
dataset.access_entries.add(access_entry)
```

Reply:
this would definitely be the right way to solve this, the issue is that the `_properties` fields do not match as there appears to be a difference in how BQ stores/returns access entries versus how they are locally constructed. The result is the local version can have extra attributes in its `_properties`.

The right long term thing to do here is to flag this issue to BQ.